lib/base/tlsstream.cpp

```diff
@@ -186,7 +186,27 @@ void TlsStream::OnEvent(int revents)
 
 					readTotal += rc;
 				}
-			} while (rc > 0 && readTotal < 64 * 1024);
+
+#ifdef I2_DEBUG /* I2_DEBUG */
+				Log(LogDebug, "TlsStream")
+					<< "Read bytes: " << rc << " Total read bytes: " << readTotal;
+#endif /* I2_DEBUG */
+				/* Limit read size. We cannot do this check inside the while loop
+				 * since below should solely check whether OpenSSL has more data
+				 * or not. */
+				if (readTotal >= 64 * 1024) {
+#ifdef I2_DEBUG /* I2_DEBUG */
+					Log(LogWarning, "TlsStream")
+						<< "Maximum read bytes exceeded: " << readTotal;
+#endif /* I2_DEBUG */
+					break;
+				}
+
+			/* Use OpenSSL's state machine here to determine whether we need
+			 * to read more data. Checking the return code of SSL_Read() is not
+			 * a reliable source unfortunately.
+			 */
+			} while (SSL_pending(m_SSL.get()));
```

Comment:
SSL_has_pending might be ever so slightly better. Other than that this looks sane enough I guess.

Reply:
Unfortunately, this doesn't exist in 1.0.2 but has been added with 1.1.0.

https://www.openssl.org/docs/man1.0.2/ssl/SSL_has_pending.html
https://www.openssl.org/docs/man1.1.0/ssl/SSL_has_pending.html